### ISSUE TITLE
NPUChain failure with single-op

### ISSUE THREAD

**Issue by ypapadop-amd** [amd/Triton-XDNA#99]
### Summary

An `NPUChain` containing a **single** op returns correct data on its first
dispatch and output shifted by exactly one `BLOCK_SIZE` on every dispatch after
that. Chains with two or more ops are unaffected.

### Reproducer

Run from the repo root (needs XRT + an npu2 device):

```python
"""NPUChain with one op: first dispatch correct, later ones shifted by BLOCK_SIZE."""
import numpy as np, torch, triton, triton.language as tl
from triton.backends.amd_triton_npu.multilaunch import NPUChain

BLOCK, N = 1024, 98304
SCRIPT = "examples/gpt2/transform_add_f32_aie2p.mlir"

@triton.jit
def add_kernel(A, B, C, n_elements: tl.constexpr, BLOCK_SIZE: tl.constexpr):
    pid = tl.program_id(0)
    off = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
    tl.store(C + off[:], tl.load(A + off[:]) + tl.load(B + off[:]))

def trial(n_ops):
    ch = NPUChain(f"repro_{n_ops}")
    z = torch.zeros(N)
    for k in range(n_ops):                       # C+0 -> tmp -> ... -> out
        ch.add(add_kernel, grid=(N // BLOCK,),
               arg_map={0: (0 if k == 0 else k + 1), 1: 1, 2: k + 2},
               args=(z, z, z, N), constexprs={"BLOCK_SIZE": BLOCK},
               transform_script=SCRIPT)
    want, out_idx = np.arange(N, dtype=np.float32), n_ops + 1
    print(f"chain with {n_ops} op(s):")
    for it in range(3):
        bufs = [want.copy()] + [np.zeros(N, np.float32) for _ in range(n_ops + 1)]
        got = ch.run(bufs, bo_key=f"k{it}", output_indices={out_idx})[out_idx]
        print(f"  dispatch {it}: out[:3]={got[:3]}  expected={want[:3]}"
              f"  {'OK' if np.array_equal(got, want) else 'WRONG'}")
    ch.close()

trial(1)
trial(2)
```

### Actual output

```
chain with 1 op(s):
  dispatch 0: out[:3]=[0. 1. 2.]           expected=[0. 1. 2.]  OK
  dispatch 1: out[:3]=[1024. 1025. 1026.]  expected=[0. 1. 2.]  WRONG
  dispatch 2: out[:3]=[1024. 1025. 1026.]  expected=[0. 1. 2.]  WRONG
chain with 2 op(s):
  dispatch 0: out[:3]=[0. 1. 2.]  expected=[0. 1. 2.]  OK
  dispatch 1: out[:3]=[0. 1. 2.]  expected=[0. 1. 2.]  OK
  dispatch 2: out[:3]=[0. 1. 2.]  expected=[0. 1. 2.]  OK
```

Expected: every dispatch returns the input unchanged (adding zero).

### Notes

- The shift is exactly `BLOCK_SIZE`; `out[i] == want[i + 1024]`.
- Independent of the buffer objects: it reproduces with a fresh `bo_key` and
  fresh host arrays on every dispatch, so it is not BO reuse or staging.
- **Uniform test data hides it** — filling the operands with a constant gives
  the correct answer no matter how the reads are offset. Non-uniform data (e.g.
  `arange`) is needed to see it.
- `examples/gpt2` is not affected: its fused MLP chain has four ops.

### Environment

- Strix Halo (Ryzen AI MAX+ 395), npu2 / `aie2p`
- XRT 2.21.75, amdxdna 2.21.0_20260602, firmware 1.1.2.65
- triton-xdna built from source at 2941882, mlir-air `d0e522b`, torch 2.13.0+rocm7.2

**Comment by erwei-xilinx** [comment]
Reproduced and root-caused. This is not a Triton-XDNA bug, and the trigger is an **odd** number of ops rather than specifically one.

## Mechanism

Each `air.launch` becomes an `aiex.configure`, which lowers to a `load_pdi @device`. The `AIEExpandLoadPdi` pass in mlir-aie rewrites each of those into *(a)* a load of an **empty** PDI — that load is what actually resets the array — followed by *(b)* explicit `write32`/`blockwrite` configuration ops.

The firmware caches the PDI address, so loading the same PDI twice in a row is a no-op. The pass knows this and alternates the empty device as `empty_ + (index % 2)`. But `index` is the module-wide `load_pdi` index, so the alternation only holds *within* one runtime sequence. The host re-runs the whole sequence on every dispatch, so it has to hold across that boundary too:

| ops | resets per dispatch | at the dispatch boundary |
|-----|---------------------|--------------------------|
| 1 | `empty_0` | `empty_0` → `empty_0` — **cached, no reset** |
| 2 | `empty_0, empty_1` | `empty_1` → `empty_0` — ok |
| 3 | `empty_0, empty_1, empty_0` | `empty_0` → `empty_0` — **cached, no reset** |
| 4 | `empty_0, empty_1, empty_0, empty_1` | ok |

From dispatch 1 onward on an odd-length chain, the configuration lands on an array that was never reset, still holding the previous run's lock and DMA state. Hence the exactly-one-`BLOCK_SIZE` phase skew.

## Confirmation

Extending the reproducer to 3 and 4 ops matches the parity model exactly:

```
chain with 1 op(s):  dispatch 0 OK, dispatch 1 WRONG, dispatch 2 WRONG
chain with 2 op(s):  all OK
chain with 3 op(s):  dispatch 0 OK, dispatch 1 WRONG, dispatch 2 WRONG
chain with 4 op(s):  all OK
```

which is also why `examples/gpt2` is unaffected — its chain has four ops.

Two further details worth recording:

- The corruption is a race, not a clean shift. Per dispatch, 4000–6000 of the 98304 elements are wrong, scattered over ~60 of the 96 blocks, varying run to run; the wrong ones are always at the *start* of a block and always satisfy `got[i] == want[i + 1024]`.
- The aircc artifacts show the PDI set directly: the 1-op build emits only `bif_empty_0.bif`, the 2-op build emits `bif_empty_0.bif` **and** `bif_empty_1.bif`.

## Fix

Xilinx/mlir-aie#3622 ("[AIEExpandLoadPdi] Keep the empty-PDI reset alternating across dispatches") appends one reset of the opposite parity to any sequence that would otherwise end on the parity it started with. That PR was opened for a different consumer — an mlir-air SmolVLA connector GEMM — so this issue is an independent second reproducer of the same defect.

Rebuilding this reproducer against an mlir-aie carrying that patch, both the 1-op and 3-op chains are correct on every dispatch:

```
chain with 1 op(s):  dispatch 0 OK, dispatch 1 OK, dispatch 2 OK
chain with 3 op(s):  dispatch 0 OK, dispatch 1 OK, dispatch 2 OK
```

No change is needed in Triton-XDNA. Once #3622 merges and an mlir-air release pins an mlir-aie containing it via the `[aie]` extra, bumping `utils/mlir-air-hash.txt` picks it up. Until then the only workaround is padding odd chains to an even op count, which I would not recommend baking in.

Environment for the above: Strix Halo, npu2/aie2p, triton-xdna at 2941882, same stack as the report.
